### PR TITLE
Add command counting visitor

### PR DESCRIPTION
**Pull request by rfilipczak**

### CODE REVIEW

File: src/main/java/edu/kis/powp/jobs2d/command/ComplexCommand.java [hunk 1]
```diff
@@ -46,6 +47,12 @@ public ComplexCommand clone() {
         return complexCommand;
     }
 
+    @Override
+    public void accept(CommandVisitor commandVisitor) {
+        commandVisitor.visitComplexCommand(this);
+        this.iterator().forEachRemaining(command -> command.accept(commandVisitor));
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/ComplexCommand.java]: remove this line

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandCountingVisitor.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+public class CommandCountingVisitor implements CommandVisitor {
+
+    private int complexCommandCount = 0;
+    private int operateToCommandCount = 0;
+    private int setPositionCommandCount = 0;
+
+    @Override
+    public void visitComplexCommand(ICompoundCommand complexCommand) {
+        complexCommandCount++;
+    }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandCountingVisitor.java]: here visitation of subcommands is required

**Review by kowallus — CHANGES_REQUESTED** [review]
please fix following issues

File: src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java [hunk 3]
```diff
@@ -59,6 +60,12 @@ public ICompoundCommand clone() {
                 return compoundCommand;
             }
 
+            @Override
+            public void accept(CommandVisitor commandVisitor) {
+                commandVisitor.visitComplexCommand(this);
+                this.iterator().forEachRemaining(command -> command.accept(commandVisitor));
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java]: iteration should be managed by visitors

File: src/main/java/edu/kis/powp/jobs2d/command/ComplexCommand.java [hunk 4]
```diff
@@ -46,6 +47,12 @@ public ComplexCommand clone() {
         return complexCommand;
     }
 
+    @Override
+    public void accept(CommandVisitor commandVisitor) {
+        commandVisitor.visitComplexCommand(this);
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/ComplexCommand.java:52]: overriding accept from ICompoudCommand seems premature

**Review by kowallus — CHANGES_REQUESTED** [review]
Please fix iteration responsibility issue

File: src/main/java/edu/kis/powp/jobs2d/command/ComplexCommand.java (hunk 1 above)
Comment (rfilipczak) [reply src/main/java/edu/kis/powp/jobs2d/command/ComplexCommand.java]: Done

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandCountingVisitor.java (hunk 2 above)
Comment (rfilipczak) [reply src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandCountingVisitor.java]: Done

File: src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java (hunk 3 above)
Comment (rfilipczak) [reply src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java]: Done

**Review by kowallus — APPROVED** [review]

**Comment by kowallus** [comment]
Note: counting visitor implementation should be stateless to provide ease of use and reliability.

**Pull request merged**